docs/library/esp32.rst

```diff
@@ -153,6 +153,133 @@ Constants
+        keyword to any input Pin
+      - ``mode_low``: set to either ``PCNT.HOLD`` or ``PCNT.REVERSE`` to
+        either suspend counting or reverse the direction of the counter (i.e.,
+        ``PCNT.INCREMENT`` behaves as ``PCNT.DECREMENT`` and vice versa)
+        when ``mode_pin`` is low
+      - ``mode_high``: as ``mode_low`` but for the behaviour when ``mode_pin``
+        is high
+      - ``filter``: set to a value 1..1023, in ticks of the 80MHz clock, to
+        enable the pulse width filter
+      - ``minimum``: set to the minimum level of the counter value when
+        decrementing (-32768..-1) or 0 to disable
+      - ``maximum``: set to the maximum level of the counter value when
+        incrementing (1..32767) or 0 to disable
+      - ``threshold0``: sets the counter value for the
+        ``PCNT.IRQ_THRESHOLD0`` event (see ``irq`` method)
+      - ``threshold1``: sets the counter value for the
+        ``PCNT.IRQ_THRESHOLD1`` event (see ``irq`` method)
+      - ``reset``: if ``True``, reset the counter value
+
+    The hardware initialisation is done in stages and so some of the keyword
+    arguments can be used in groups or in isolation to partially reconfigure a
+    unit:
+
+      - the ``pin`` keyword (optionally combined with ``mode_pin``) can be used
+        to change just the bound pin(s)
+      - ``rising`, ``falling``, ``mode_low`` and ``mode_high`` can be used
+        (singly or together) to change the counting logic - omitted keywords
+        use their default (``PCNT.IGNORE`` or ``PCNT.NORMAL``)
+      - ``filter`` can be used to change only the pulse width filter (with 0
+        disabling it)
+      - each of ``minimum``, ``maximum``, ``threshold0`` and ``threshold1`` can
+        be used to change these limit/event values individually; however,
+        setting any will reset the counter to zero (i.e., they imply
+        ``value=0``)
+
+    Each pulse counter unit supports two channels, 0 and 1, each able to
+    monitor different pins with different counting logic but updating the same
+    counter value. Use ``channel=1`` with the ``pin``, ``rising``, ``falling``,
+    ``mode_pin``, ``mode_low`` and ``mode_high`` keywords to configure the
+    second channel.
+
+    The second channel can be used to configure 4X quadrature decoding with a
+    single counter unit::
+
+        pin_a = Pin(2, Pin.INPUT, pull=Pin.PULL_UP)
+        pin_b = Pin(3, Pin.INPUT, pull=Pin.PULL_UP)
+        rotary = PCNT(0, minimum=-32000, maximum=32000)
+        rotary.init(channel=0, pin=pin_a, falling=PCNT.INCREMENT, rising=PCNT.DECREMENT, mode_pin=pin_b, mode_low=PCNT.REVERSE)
+        rotary.init(channel=1, pin=pin_b, falling=PCNT.DECREMENT, rising=PCNT.INCREMENT, mode_pin=pin_a, mode_low=PCNT.REVERSE)
+        rotary.start()
+
+.. method:: PCNT.value([reset])
+
+    Call this method with no arguments to return the current counter value.
+
+    If the optional *reset* argument is set to ``True`` then the counter is
```

Comment:
This feels inconsistent with other machine APIs: if a method is acting like a getter and setter then the setter should take the value to set it to, which in this case would be 0, not `True`.

Reply:
I think this was the primary reason why I originally went with `.value(0)` for resetting the counter, as it allowed for a (near-)atomic read-and-reset while fitting with the setter convention.

Reply:
The intention with the change is that having the ability to set the value but only allowing the set value to be zero is a bit weird. What I wanted to convey is that the "action" is reset/clear, not "set to value". It would probably make sense if we required this to be a kwarg-only... e.g. `pcnt.value(reset=True)`.

Either way, I don't care that much, I've reverted it back to just value-that-must-be-zero. (And same for init).

Reply:
`reset=True` as a keyword-only argument makes sense to me. You're right that making `.value()` a combined getter/setter feels weird when you can only set it to `0` – honestly, I was never particularly comfortable with that bit of my API. I think that having `.value(0)` return the previous value is already stretching the setter convention a bit thin…

Reply:
In the general case .value(some_non_zero_value) is possible. See 

https://github.com/IhorNehrutsa/micropython/blob/7b82d997eb930dd552b7701fddff4c7b38f0890c/ports/esp32/machine_encoder.c#L235-L253

Reply:
> In the general case .value(some_non_zero_value) is possible. 

Yes, that is still supported on `machine.Counter` and `machine.Encoder`. This limitation only applies to the underlying `esp32.PCNT` class which doesn't manage any internal counter state.